### ISSUE TITLE
afterNextRender documentation lacks details and limitations

### ISSUE THREAD

**Issue by kemsky** [angular/angular#55870]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

We are using `MessageChannel.portX.postMessage` and `Promise.resolve().then` to schedule tasks and micro-tasks. It allows to schedule code to execute after render (at least it used to work) and after certain callbacks. Also, our code is aware of zones, so it will run angular tasks in angular zone and root tasks in root zone.

I've tried to replace `MessageChannel` hack with `afterNextRender` which in theory should be a drop-in replacement but it quickly became obvious that it will not work because it requires injection context or explicitly provided injector which is extremely inconvenient and leads to errors that are difficult to reason about.

### Please provide the exception or error you saw

```true
NG0203 - missing context, then other non-angular errors due to unexpected context because I provided app module injector.
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 17.1.2
Node: 20.13.1
Package Manager: yarn 3.5.1
OS: linux x64

Angular: 17.1.2
... animations, cdk, cli, common, compiler, compiler-cli, core
... elements, forms, language-service, material
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1701.2
@angular-devkit/build-angular   17.1.2
@angular-devkit/core            17.1.2
@angular-devkit/schematics      17.1.2
@schematics/angular             17.1.2
rxjs                            7.8.1
typescript                      5.3.3
zone.js                         0.14.3
```

**Comment by alxhub** [comment]
Hi,

Can you provide some context on what exactly you think is missing, besides documentation of the requirement for injection context?

**Comment by JoostK** [comment]
@benelliott if you can replace the observable with a signal, [the new `afterRenderEffect`](https://next.angular.dev/api/core/afterRenderEffect#afterRenderEffect_1) should help make this more ergonomic:

```ts
afterRenderEffect({
  write: () => this.doSomethingToTheDom(),
});
```

Here, any signal that is read in `this.doSomethingToTheDom()` will be considered a dependency of the DOM state, retriggering the DOM syncing on change, scheduled at an appropriate time w.r.t. rendering/change detection.

**Comment by JoostK** [comment]
Or with an observable, I suppose something along the lines of the following could work:

```ts
const signal = toSignal(myObservable$);

afterRenderEffect({
  write: () => this.doSomethingToTheDom(signal()),
});
```

If done in the constructor this follows the lifecycle of the component, destroying the observable subscription (that drives the signal) and the render effect as expected.

**Comment by kemsky** [comment]
@JoostK, the one and only thing that `afterNextRender`  is supposed to do is to schedule callback to execute after next render (or render phase), it should not matter how, when, where it was triggered. Current implementation for some unknown reason was optimized to work in component constructor, it requires context (implicit or explicit). It is not documented what context can be supplied to this function, why and what will be consequences. Also, it is not specified how this works related to zones. `afterNextRender`  simply can not be used in production code until these questions get answered.

Adding signals to this mess will only further complicate logic, introduce new hidden effects, timing issues and so on.

**Comment by kemsky** [comment]
`injector` property of `[AfterRenderOptions](https://next.angular.dev/api/core/AfterRenderOptions)` is marked as deprecated, so now I have no idea how this function is even supposed to be used.

**Comment by JeanMeche** [comment]
We have a pending PR that fixes this : #57595

**Comment by JeanMeche** [comment]
Currently the injector param states `If this is not provided, the current injection context will be used instead`. 

I think this is clear enough.
